skorch/callbacks/training.py

```diff
@@ -273,3 +278,145 @@ def _sink(self, text, verbose):
         #  We do not want to be affected by verbosity if sink is not print
         if (self.sink is not print) or verbose:
             self.sink(text)
+
+
+
+class ParamMapper(Callback):
+    """Map arbitrary functions over module parameters filtered by pattern
+    matching. In the simplest case the function is only applied once at
+    the beginning of a given epoch (at ``on_epoch_begin``) but more complex
+    execution schemes (e.g. periodic application) are possible using
+    ``at`` and ``scheduler``.
+
+    Examples
+    --------
+    Initialize a layer on first epoch before the first training step:
+
+    >>> init = partial(torch.nn.init.uniform_, a=0, b=1)
+    >>> cb = ParamMapper('linear*.weight', at=1, fn=init)
+    >>> net = Net(myModule, callbacks=[cb])
+
+    Reset layer initialization if train loss reaches a certain value:
+
+    >>> at = lambda net: net.history[-1, 'train_loss'] > 1
+    >>> init = partial(torch.nn.init.uniform_, a=0, b=1)
+    >>> cb = ParamMapper('linear0.weight', at=at, fn=init)
+    >>> net = Net(myModule, callbacks=[cb])
+
+    Periodically freeze and unfreeze all embedding layers:
+
+    >>> def my_sched(net, _at, _fn):
+    ...    if len(net.history) % 2 == 0:
+    ...        return skorch.utils.freeze_parameter
+    ...    else:
+    ...        return skorch.utils.unfreeze_parameter
+    >>> cb = ParamMapper('embedding*.weight', schedule=my_sched)
+    >>> net = Net(myModule, callbacks=[cb])
+
+    Parameters
+    ----------
+    patterns : str or callable or list
+      The pattern(s) to match parameter names against. Patterns are
+      UNIX globbing patterns as understood by :func:`~fnmatch.fnmatch`.
+      Patterns can also be callables which will get called with the
+      parameter name and are regarded as a match when the callable
+      returns a truthy value.
+
+      Example: ``'linear*.weight'`` or ``['linear0.*', 'linear1.bias']``
+            or ``lambda name: name.startswith('linear')``.
+
+    fn : function
+      The function to apply to each parameter separately.
+
+    at : int or callable
+      In case you specify an integer it represents the epoch number the
+      function ``fn`` is applied to the parameters, in case ``at`` is
+      a function it will receive ``net`` as parameter and the function
+      is applied to the parameter once ``at`` returns ``True``.
+
+    schedule : callable or None
+      If specified it determines how ``fn`` and ``at`` are used.
+      This function receives ``net``, ``at`` and ``fn`` and returns
+      the function to be applied on the matched parameters.
```

Comment:
I have to admit I don't see what use case you have in mind for `schedule`, especially since the example above ("periodically freeze and unfreeze ...") could also be achieved by `at`.

Reply:
I don't see how it can be achieved with `at`. You'd need to create two `ParamMapper` to solve this with `at`, e.g.

    ParamMapper(fn=freeze, at=lambda net: len(net.history) % 2 == 0)
    ParamMapper(fn=unfreeze, at=lambda net: len(net.history) % 2 == 1)

which is not very grid search-friendly.

Reply:
My point is that I would like to see an example that uses `at` and `fn`, because otherwise, `schedule` seems to be pretty orthogonal to those two. If there is no conceivable case for that, we have to wonder why those parameters are passed. Or maybe the schedule functionality needs to be in another class, so that you have either `at` and `fn`, or `schedule`, but not both.

Reply:
Since `schedule` is the generalisation of the static `at`/`fn` case I think it makes sense to have it here. While I can see that it is possible to keep the implementation but separate the current implementation in `StaticParamMapper` (supports `at`/`fn`) and `PeriodicParamMapper` (supports `schedule`), I don't think it adds much clarity.

The parameters on the other hand are, indeed, quite useless at it seems. I suggest to remove them. We could possibly provide the callback instance in addition to `net` but at least for the default implementation of the schedule this is not necessary (as it is a method and can access `self.at` anyway). The new signature would then be `schedule(net: NeuralNet) -> callable`.

Reply:
Okay, this looks like a good compromise. Make sure to mention in the `schedule` docstring that depending on the input, `schedule` may supersede `at` and `fn`.